src/main/java/org/dataone/hashstore/hashfs/HashFileStore.java

```diff
@@ -46,91 +54,162 @@ public HashFileStore(int depth, int width, String algorithm, String storeDirecto
+        if (object == null) {
+            throw new NullPointerException("Invalid input stream, data is null.");
+        }
+        // pid cannot be empty or null
+        if (pid == null || pid == "") {
+            // TODO: Log failure - include signature values
+            throw new IllegalArgumentException("The pid cannot be null or empty");
+        }
+
+        // Checksum cannot be empty if checksumAlgorithm is passed
+        if (checksumAlgorithm != null & checksum == "") {
+            // TODO: Log failure - include signature values
+            throw new IllegalArgumentException("Checksum cannot be null or empty");
+        }
         // Cannot generate additional algorithm if it is not supported
         boolean algorithmSupported = this.hsil.validateAlgorithm(additionalAlgorithm);
+        boolean checksumAlgorithmSupported = this.hsil.validateAlgorithm(checksumAlgorithm);
         if (!algorithmSupported) {
             // TODO: Log failure - include signature values
             throw new IllegalArgumentException(
-                    "Algorithm not supported. Supported algorithms: " + this.hsil.supportedHashAlgorithms);
+                    "Additional algorithm not supported - unable to generate additional hex digest value. additionalAlgorithm: "
+                            + additionalAlgorithm + ". Supported algorithms: "
+                            + this.hsil.supportedHashAlgorithms);
+        }
+        if (!checksumAlgorithmSupported) {
+            // TODO: Log failure - include signature values
+            throw new IllegalArgumentException(
+                    "Checksum algorithm not supported - cannot be used to validate object. checksumAlgorithm: "
+                            + checksumAlgorithm + ". Supported algorithms: "
+                            + this.hsil.supportedHashAlgorithms);
+        }
+
+        // Gather HashAddress elements and prepare object permanent address
+        String objAuthorityId = this.hsil.getHexDigest(pid, this.objectStoreAlgorithm);
+        String objShardString = this.hsil.shard(directoryDepth, directoryWidth, objAuthorityId);
+        String objAbsolutePathString = this.objectStoreDirectory.toString() + objShardString;
+        File objHashAddress = new File(objAbsolutePathString);
+        // If file (pid hash) exists, reject request immediately
+        if (objHashAddress.exists()) {
+            throw new FileAlreadyExistsException("File already exists for pid: " + pid);
         }
 
         // Generate tmp file and write to it
         File tmpFile = this.hsil.generateTmpFile("tmp", this.tmpFileDirectory.toFile());
         Map<String, String> hexDigests = this.hsil.writeToTmpFileAndGenerateChecksums(tmpFile, object,
                 additionalAlgorithm);
 
-        // Validate object if algorithm and checksum is passed
+        // Validate object if checksum and checksum algorithm is passed
         if (additionalAlgorithm != null && checksum != null) {
```

Comment:
This line confuses me.  Is the `checksum` is for  `additionalAlgorithm` or `checksumAlgorithm`?

Reply:
It was a mistake, I have addressed it in my latest commits. It should be for checksumAlgorithm.